pytorch_lightning/plugins/training_type/training_type_plugin.py

```diff
@@ -91,26 +92,53 @@ def is_global_zero(self) -> bool:
         """Whether the current process is the rank zero process not only on the local node, but for all nodes."""
 
     @abstractmethod
-    def reduce(self, tensor: Union[torch.Tensor, Any], *args: Any, **kwargs: Any) -> Union[torch.Tensor, Any]:
+    def reduce(
+        self,
+        tensor: Union[torch.Tensor, Any],
+        group: Optional[Any] = None,
+        reduce_op: Optional[Union[ReduceOp, str]] = "mean",
+    ) -> Union[torch.Tensor, Any]:
         """Reduces the given tensor (e.g. across GPUs/processes).
 
         Args:
             tensor: the tensor to sync and reduce
+            group: the process group to reduce
+            reduce_op: the reduction operation. Defaults to 'mean'.
+                Can also be a string 'sum' or ReduceOp.
             *args: plugin-specific positional arguments
             **kwargs: plugin-specific keyword arguments
         """
 
     @abstractmethod
     def barrier(self, name: Optional[str] = None) -> None:
-        """Forces all possibly joined processes to wait for each other."""
+        """Synchronizes all processes which blocks processes until the whole group enters this function.
+
+        Args:
+            name: a str pass into barrier. Only torch xla respect this param
+        """
 
     @abstractmethod
-    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
-        """Broadcasts an object to all processes."""
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        """Broadcasts an object to all processes.
+
+        Args:
+            obj: the object to broadcast
+            src: source rank.
+        """
 
     @abstractmethod
-    def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_grads: bool = False) -> torch.Tensor:
-        """Perform a all_gather on all processes."""
+    def all_gather(
+        self, tensor: torch.Tensor, group: Optional[Any] = None, sync_grads: bool = False
+    ) -> Union[List[torch.Tensor], torch.Tensor]:
+        """Perform a all_gather on all processes.
+
+        Args:
+            tensor: the tensor to all_gather
+            group: the process group to gather results from
+            sync_grads: flag that allows users to synchronize gradients for all_gather op
+
+        Returns: a tensor (torch distributed) or a list of tensor (horovod)
```

Comment:
- i don't think we should list the particular implementations return type specifics in the base class
- for horovod, returning a list depends on whether we're calling `allgather` or `allgather_object`: https://horovod.readthedocs.io/en/stable/_modules/horovod/torch/mpi_ops.html#allgather

since the input here is a tensor, we should be calling `horovod.torch.allgather` which returns a tensor

Reply:
>i don't think we should list the particular implementations return type specifics in the base class
It's one of the comments I got from #9414, ppl like having explains when which will be returned. Should we remove it instead?

>for horovod, returning a list depends on whether we're calling allgather or allgather_object: https://horovod.readthedocs.io/en/stable/_modules/horovod/torch/mpi_ops.html#allgather since the input here is a tensor, we should be calling horovod.torch.allgather which returns a tensor

For some reason, current horovod implementation calls allgather but then convert it to a list of tensor. @awaelchli @justusschock do you guys know why?

Reply:
is this a bug with the lightning horovod implementation? https://github.com/PyTorchLightning/pytorch-lightning/blob/8fcdcb598bbf0fbe6cda92415d1d69eb1e2d0aa0/pytorch_lightning/plugins/training_type/horovod.py#L174-L188

the typehint doesn't match what's being returned

Reply:
@four4fish what breaks if we change the horovod implementation to directly return hvd all gather instead of a list?

Reply:
I think so, if user call lightningModule.all_gather() and the TTP is horovod, the return will be different.

Reply:
IMO this is a bugfix. We can mark that this is a breaking change, but this is inconsistent with the API/typehint/docs elsewhere. calling allgather with a tensor should do what horovod.torch does now

git blame points to #5714 but it's likely this was a bug even before

Reply:
Horovod docs: https://horovod.readthedocs.io/en/stable/api.html#horovod.torch.allgather
they return a tensor just like torch.distributed. Us converting to a list is a bit silly, we should remove the list conversion, I don't see why it is there.